In [1]:
import pandas as pd

# Data Gathering

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

# Assessing

`Visually`

In [3]:
patients

patient_id assigned_sex  given_name       surname  \
0             1       female         Zoe       Wellish   
1             2       female      Pamela          Hill   
2             3         male         Jae        Debord   
3             4         male        Liêm          Phan   
4             5         male         Tim       Neudorf   
5             6         male      Rafael         Costa   
6             7       female        Mary         Adams   
7             8       female      Xiuxiu         Chang   
8             9         male       Dsvid    Gustafsson   
9            10       female      Sophie       Cabrera   
10           11       female       Sandy    Gunnarsson   
11           12         male   Abdul-Nur           Isa   
12           13         male  Omeokachie      Ibeamaka   
13           14       female    Anenechi         Chidi   
14           15       female        Asia       Woźniak   
15           16         male       Søren          Lund   
16           17       female         Tám          Liễu   
17           18       female     Roxanne     Andreyeva   
18           19         male     William         Oates   
19           20         male         Zak         Kelly   
20           21       female       Sofia       Karlsen   
21           22         male      Samúel  Guðbrandsson   
22           23         male      Manchu            Su   
23           24         male       Lovre         Galić   
24           25         male       Jakob      Jakobsen   
25           26         male      Gregor          Bole   
26           27       female        Ella          Lund   
27           28         male      Joseph        Tucker   
28           29         male      Robert          Wolf   
29           30         male        Jake      Jakobsen   
..          ...          ...         ...           ...   
473         474       female        Kate     Wilkinson   
474         475       female   Esperanza      Labrosse   
475         476         male       Malik       Vaneker   
476         477       female       Berta    Napolitani   
477         478         male     Juliusz      Majewski   
478         479       female      Edelma   Villalpando   
479         480         male        Tapa   Arsanukayev   
480         481         male      Nasser       Mansour   
481         482         male     Michael    Kristensen   
482         483         male       Diogo         Souza   
483         484       female       Angel         Grant   
484         485         male     Placido       Udinesi   
485         486         male      Trifon      Izmailov   
486         487         male      Samuel          Blix   
487         488         male        Ivar       Löfgren   
488         489         male        Mika    Martinsson   
489         490       female     Jasmine         Sykes   
490         491         male     Jackson       Addison   
491         492       female     Vanessa      Ferguson   
492         493         male       Poldi           Tar   
493         494       female         Fen          Chin   
494         495       female      Sirkka     Piirainen   
495         496         male      Hajime       Tsukada   
496         497         male   Alexander        Hueber   
497         498         male    Masataka      Murakami   
498         499         male     Mustafa     Lindström   
499         500         male       Ruman       Bisliev   
500         501       female       Jinke     de Keizer   
501         502       female     Chidalu   Onyekaozulu   
502         503         male         Pat       Gersten   

                        address               city       state  zip_code  \
0          576 Brown Bear Drive  Rancho California  California   92390.0   
1     2370 University Hill Road          Armstrong    Illinois   61812.0   
2         1493 Poling Farm Road               York    Nebraska   68467.0   
3           2335 Webster Street         Woodbridge          NJ    7095.0   
4          1428 Turkey Pen L

In [4]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [5]:
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

`Progmatically`

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [8]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [9]:
patients['surname'].value_counts()

Doe              6
Jakobsen         3
Taylor           3
Liễu             2
Ogochukwu        2
Tucker           2
Silva            2
Correia          2
Bùi              2
Cabrera          2
Parker           2
Gersten          2
Woźniak          2
Tạ               2
Kowalczyk        2
Kadyrov          2
Cindrić          2
Lương            2
Johnson          2
Grímsdóttir      2
Lund             2
Nilsen           2
Schiavone        2
Berg             2
Lâm              2
Collins          2
Batukayev        2
Dratchev         2
Hueber           2
Souza            2
                ..
Rokavc           1
Teichelmann      1
Traustadóttir    1
Bisliev          1
Hansen           1
Mathiesen        1
Dvořák          1
Nebay            1
Iwata            1
Johnsrud         1
Navrátil         1
Nowakowski       1
Žibrik           1
Selassie         1
Kelly            1
Lindgren         1
Grant            1
Vũ               1
Amanuel          1
Jackson          1
Mensa            1
Rahal       

In [10]:
patients['address'].value_counts()

123 Main Street              6
648 Old Dear Lane            2
2476 Fulton Street           2
2778 North Avenue            2
3259 Roy Alley               1
2356 Myra Street             1
2531 Eastland Avenue         1
2103 Edington Drive          1
1953 Rhapsody Street         1
4310 Johnson Street          1
4220 Simpson Square          1
3084 Blue Spruce Lane        1
2754 Westwood Avenue         1
2246 Pheasant Ridge Road     1
1690 Fannie Street           1
4018 Poplar Avenue           1
1233 Liberty Avenue          1
4894 Chandler Hollow Road    1
1493 Randolph Street         1
4120 Thompson Drive          1
666 Whiteman Street          1
3743 Wescam Court            1
1333 Comfort Court           1
1703 Everette Alley          1
3652 Boone Crockett Lane     1
230 Lady Bug Drive           1
3391 Marcus Street           1
3539 Bottom Lane             1
1717 Vineyard Drive          1
4643 Reeves Street           1
                            ..
1619 Melm Street             1
2403 Rip

In [11]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [12]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [64]:
patients.state.value_counts()

California    36
TX            32
New York      25
CA            24
NY            22
MA            22
PA            18
GA            15
Illinois      14
OH            14
LA            13
OK            13
MI            13
Florida       13
NJ            12
VA            11
IL            10
WI            10
MS            10
TN             9
AL             9
IN             9
MN             9
FL             9
NC             8
WA             8
KY             8
MO             7
ID             6
NV             6
KS             6
SC             5
IA             5
CT             5
ME             4
AZ             4
Nebraska       4
ND             4
AR             4
CO             4
RI             4
SD             3
DE             3
WV             3
MD             3
OR             3
VT             2
NE             2
MT             2
DC             2
WY             1
NM             1
AK             1
NH             1
Name: state, dtype: int64

In [13]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [14]:
all_columns = pd.Series(list(patients)+list(adverse_reactions)+list(treatments))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
17    given_name
18       surname
dtype: object

In [15]:
treatments.sample(10)

given_name      surname    auralin    novodra  hba1c_start  hba1c_end  \
13       gregor         bole          -  47u - 45u         7.61       7.16   
176       andré       nordin  40u - 47u          -         7.60       7.27   
18   addolorata     lombardi          -  49u - 46u         7.75       7.33   
26       suhaim        rahal          -  49u - 47u         7.94       7.50   
32        laura  ehrlichmann          -  43u - 40u         7.95       7.46   
275      albina     zetticci  45u - 51u          -         7.93       7.73   
140       jesse        luoma          -  39u - 37u         7.72       7.35   
250        chen          yao          -  56u - 57u         7.90       7.51   
106       sofia      karlsen          -  35u - 29u         7.62       7.15   
257    mathilde     nørgaard          -  27u - 28u         8.50       8.10   

     hba1c_change  
13           0.95  
176          0.33  
18            NaN  
26           0.94  
32           0.99  
275          0.20  
140           NaN  
250          0.39  
106          0.97  
257          0.90

### Quality issues

#### Patients Table

* Zip code is of 4 digits sometimes
* Typo of given_name at the 8th index
* Height is inaccurate at 4th index
* Few datatypes are wrong(birthday,sex,zipcode)
* Missing info in some columns(address,city,state,zipcode,country,contact)
* Multiple phone number format
* John Due Duplicated

#### Treatments Table

* hb1ac change missing in many values
* The letter u present in staring and ending of columns auralin and novodra
* Lowercase given_name and surname
* Erroneous datatypes(auralin and novodra)
* Some hb1ac diffrence is calculated wrong
* Missing values(280 instead of 350)

#### Adverse Reaction Table

* Lowercase given_name and surname


### Tideness Issues:

* In the **treatments** table there should be one variable drug type 
* There should be only 2 tables in all (adverse reaction and treatment should combine as they're related)
* Contacts column in **treatments** table should be split into two different columns
* given_name and surname columns in **patients** table duplicated in **treatments** and **adverse_reactions*** tables

# Cleaning Data 

In [16]:
treatments_clean=treatments.copy()
patients_clean=patients.copy()
adverse_reactions_clean=adverse_reactions.copy()

In [17]:
patients_clean=patients.copy()


## Missing Data 

### Missing `Treatments` table values

#### Define

`Treatments Table`- We will concatenate the missing values(stored in treatments_cut.scv) with the treatments_copy dataframe.

#### Code

In [18]:
treat_miss=pd.read_csv("treatments_cut.csv")
treatments_clean=pd.concat([treatments_clean,treat_miss],ignore_index=True)


#### Test

In [19]:
treatments_clean.shape

(350, 7)

### Missing `hba1c values` and correcting the wrong `hba1c values`  

#### Define:

Will calculate hba1c values for all columns by  *hba1c_end - hba1c_start*

#### Code

In [20]:
treatments_clean['hba1c_change']=treatments_clean['hba1c_end']-treatments_clean['hba1c_start']

#### Test

In [21]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    350 non-null float64
dtypes: float64(3), object(4)
memory usage: 19.2+ KB


## Tidiness Issues

#### Define
Contacts column in `Patients Table` need to be split into email and phone number

#### Code

In [22]:
patients_clean['email']=patients_clean['contact'].str.extract(r'([A-Za-z]+@[a-z]+\.[a-z]+)')


In [23]:
patients_clean['phone']=patients_clean['contact'].str.extract(r'([\+\d]*[\s(]*[\d]{3}[\s\)\-]+\d{3}[\s\)\-]\d{3})')

In [24]:
patients_clean.drop(columns="contact",axis=1,inplace=True)

#### Test

In [25]:
patients_clean.sample(10)

patient_id assigned_sex given_name    surname                  address  \
495         496         male     Hajime    Tsukada        4111 Thunder Road   
139         140       female    Novalie       Berg         1275 Goldie Lane   
278         279       female   Generosa      Cabán                      NaN   
291         292         male       Enco     Žibrik         4792 Maud Street   
119         120       female      Chaja       Bouw        2808 Circle Drive   
179         180         male    Dominik  Grunewald      3574 Park Boulevard   
121         122         male    Frydryk    Adamski  1845 Saint Marys Avenue   
490         491         male    Jackson    Addison       1160 Taylor Street   
96           97       female      Nasim      Salib        3161 Fantages Way   
154         155       female       Maya   Isaksson       1333 Comfort Court   

               city     state  zip_code        country   birthdate  weight  \
495       San Mateo        CA   94403.0  United States    9/5/1972   168.1   
139      Cincinnati        OH   45202.0  United States   4/22/1973   156.9   
278             NaN       NaN       NaN            NaN  12/16/1962   124.3   
291  Camden Wyoming        DE   19934.0  United States   1/25/1983   203.9   
119         Houston        TX   77063.0  United States   9/16/1979   211.9   
179    Marshalltown        IA   50158.0  United States  11/20/1935   143.0   
121        Syracuse  New York   13221.0  United States  10/12/1941   230.8   
490    New Rochelle  New York   10801.0  United States   5/29/1953   192.7   
96     Presque Isle        ME    4769.0  United States   7/30/1971   211.4   
154         Madison        WI   53711.0  United States   7/14/1959   137.7   

     height   bmi                         email             phone  
495      66  27.1      HajimeTsukada@dayrep.com       650-570-489  
139      61  29.6         NovalieBerg@gustr.com  +1 (513) 383-051  
278      69  18.4                           NaN               NaN  
291      74  26.2      EncoZibrik@superrito.com       302-698-205  
119      70  30.4          ChajaBouw@dayrep.com       832-457-662  
179      71  19.9     DominikGrunewald@cuvox.de       641-753-567  
121      76  28.1    FrydrykAdamski@armyspy.com     1 315 869 440  
490      69  28.5    JacksonAddison@armyspy.com       914-636-930  
96       65  35.2  NasimSumaiyaSalib@einrot.com       207-768-047  
154      67  21.6   MayaIsaksson@jourrapide.com     1 608 277 014

#### Define

Three variables in two columns in **`treatments_clean`** df (treatment, start dose and end dose).Will use the pandas melt function and str.split() to make auralin novodra into a column named `drug_type` and other columns  `start_dose` and `end_dose` 

#### Code:

In [26]:
a=pd.melt(treatments_clean,id_vars=["given_name","surname","hba1c_start","hba1c_end","hba1c_change"],var_name="drug_type",value_name="Dosage")

In [27]:
a["Dosage"]=   a["Dosage"].str.extract(r'([\d]{2}[u]\s[\-]\s[\d]{2}[u])')

In [28]:
a.dropna(subset=["Dosage"],inplace=True)

In [29]:
a["start_dose"],a["end_dose"]=a.Dosage.str.split("-",1).str

In [30]:
a.drop(columns="Dosage",axis=1,inplace=True)

In [31]:
treatments_clean=a.copy()

#### Test

In [32]:
treatments_clean.sample(5)

given_name   surname  hba1c_start  hba1c_end  hba1c_change drug_type  \
621        leo    vieira         7.74       7.36         -0.38   novodra   
457   flaviano  endrizzi         7.93       7.56         -0.37   novodra   
29      eugene   mironov         7.81       7.48         -0.33   auralin   
174    mustafa    adonay         7.79       7.34         -0.45   auralin   
266     ursula     freud         7.75       7.46         -0.29   auralin   

    start_dose end_dose  
621       30u       33u  
457       46u       43u  
29        42u       49u  
174       58u       69u  
266       42u       54u

#### Define

**`Adverse Reactions`** should be a part of **`Treatments`** Table. We will merge them using merge function joining on `given_name` and `surname`

#### Code

In [33]:
treatments_clean=pd.merge(treatments_clean,adverse_reactions_clean,how="left", on=['given_name','surname'])

#### Test

In [34]:
treatments_clean.sample(5)

given_name    surname  hba1c_start  hba1c_end  hba1c_change drug_type  \
293       anja     hueber         8.82       8.36         -0.46   novodra   
48      marius     hansen         7.97       7.53         -0.44   auralin   
340    satsita  batukayev         7.63       7.25         -0.38   novodra   
104    vanessa   ferguson         7.97       7.58         -0.39   auralin   
292      eemil      laine         7.89       7.53         -0.36   novodra   

    start_dose end_dose adverse_reaction  
293       31u       32u              NaN  
48        41u       47u              NaN  
340       42u       42u              NaN  
104       39u       50u              NaN  
292       30u       29u              NaN

#### Define:

Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables.We will replace that column with patient_id

#### Code:

In [35]:
b=patients_clean["surname"].to_frame()
c=patients_clean["given_name"].to_frame()
d=patients_clean["patient_id"].to_frame()

In [36]:
pat_id=pd.concat([b,c,d],axis=1)
pat_id.given_name=pat_id.given_name.str.lower()
pat_id.surname=pat_id.surname.str.lower()

In [37]:
treatments_clean=pd.merge(pat_id,treatments_clean,on=["given_name","surname"])


In [38]:
treatments_clean.drop(["given_name","surname"],axis=1,inplace=True)

#### Test

In [39]:
treatments_clean.sample()

patient_id  hba1c_start  hba1c_end  hba1c_change drug_type start_dose  \
82         126         7.66        7.3         -0.36   auralin       39u    

   end_dose adverse_reaction  
82      46u              NaN

## Quality Issues

#### Define
In `patients` table zip code is a float and not a string, zip code has 4 digits sometimes. Convert the datatype and add a 0 in the start when there are 4 digits

#### Code

In [53]:
patients_clean.zip_code=patients_clean.zip_code.astype(str).str[:-2]


#### Test

In [54]:
patients_clean.sample()

patient_id assigned_sex given_name surname                    address  \
133         134       female  Elisabeth  Dimmen  3180 Still Pastures Drive   

         city state zip_code        country  birthdate  weight  height   bmi  \
133  Columbia    SC    29210  United States  4/21/1985   199.5      65  33.2   

                        email             phone  
133  ElisabethDimmen@cuvox.de  +1 (803) 465-331

In [55]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        503 non-null object
country         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
email           491 non-null object
phone           485 non-null object
dtypes: float64(2), int64(2), object(11)
memory usage: 59.0+ KB


#### Define:
Height of Tim Neudorf(at 4th index) is wrong. Will change it from 27 to 72 in the `patients` table

#### Code:

In [57]:
patients_clean.height[4]=72

/Users/ishanarora/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


#### Test:

In [58]:
patients_clean.height[4]

72

#### Define 
In `patients` table given_name should be "david" instead of "dsvid"

#### Code

In [62]:
patients_clean.given_name[8]="David"

/Users/ishanarora/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


#### Test:

In [63]:
patients_clean.given_name[8]

'David'

#### Define:
There are  abbreviations used sometimes and full state name used other times in the state column of `patients` table.Will convert the full state name into abbreviations.

#### Code

In [67]:
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}


def abbreviate_state(patient):
    if patient['state'] in state_abbrev.keys():
        abbrev = state_abbrev[patient['state']]
        return abbrev
    else:
        return patient['state']
    
patients_clean['state'] = patients_clean.apply(abbreviate_state, axis=1)

#### Test:

In [68]:
patients_clean.state.value_counts()

CA    60
NY    47
TX    32
IL    24
FL    22
MA    22
PA    18
GA    15
OH    14
LA    13
OK    13
MI    13
NJ    12
VA    11
WI    10
MS    10
IN     9
MN     9
TN     9
AL     9
NC     8
WA     8
KY     8
MO     7
ID     6
KS     6
NV     6
NE     6
IA     5
CT     5
SC     5
RI     4
CO     4
ND     4
ME     4
AR     4
AZ     4
WV     3
SD     3
OR     3
MD     3
DE     3
MT     2
DC     2
VT     2
NH     1
NM     1
AK     1
WY     1
Name: state, dtype: int64